src/docker/ContextManager.ts

```diff
@@ -0,0 +1,201 @@
+// Consider having our own timeout for execution of any context-related Docker CLI commands.
+// The following timeout is for _starting_ the command only; in the current implementation there is no timeot
+// for command duration.
+const ContextCmdExecOptions: ExecOptions = { timeout: 5000 }
+
+const dockerConfigFile = path.join(os.homedir(), '.docker', 'config.json');
+const dockerContextsFolder = path.join(os.homedir(), '.docker', 'contexts', 'meta');
+
+const WindowsLocalPipe = 'npipe:////./pipe/docker_engine';
+const UnixLocalPipe = 'unix:///var/run/docker.sock';
+
+const defaultContext: Partial<DockerContext> = {
+    Name: 'default',
+    Description: 'Current DOCKER_HOST based configuration',
+    CreatedTime: undefined, // Not defined for contexts
+};
+
+export interface ContextManager {
+    readonly onContextChanged: Event<DockerContext>;
+    refresh(): Promise<void>;
+    getContexts(): Promise<DockerContext[]>;
+
+    inspect(actionContext: IActionContext, contextName: string): Promise<DockerContextInspection>;
+    use(actionContext: IActionContext, contextName: string): Promise<void>;
+    remove(actionContext: IActionContext, contextName: string): Promise<void>;
+}
+
+// TODO: consider a periodic refresh as a catch-all; but make sure it compares old data to new before firing a change event
+// TODO: so that non-changes don't result in everything getting refreshed
+export class DockerContextManager implements ContextManager, Disposable {
+    private readonly emitter: EventEmitter<DockerContext> = new EventEmitter<DockerContext>();
+    private readonly contextsCache: AsyncLazy<DockerContext[]>;
+    private readonly configFileWatcher: fs.FSWatcher;
+    private readonly contextFolderWatcher: fs.FSWatcher;
+    private refreshing: boolean = false;
+
+    public constructor() {
+        this.contextsCache = new AsyncLazy(async () => this.loadContexts());
+
+        /* eslint-disable @typescript-eslint/tslint/config */
+        this.configFileWatcher = fs.watch(dockerConfigFile, async () => this.refresh());
+        this.contextFolderWatcher = fs.watch(dockerContextsFolder, async () => this.refresh());
+        /* eslint-enable @typescript-eslint/tslint/config */
+    }
+
+    public dispose(): void {
+        void this.configFileWatcher?.close();
+        void this.contextFolderWatcher?.close();
+        void ext.dockerClient?.dispose();
+    }
+
+    public get onContextChanged(): Event<DockerContext> {
+        return this.emitter.event;
+    }
+
+    public async refresh(): Promise<void> {
+        if (this.refreshing) {
+            return;
+        }
+
+        try {
+            this.refreshing = true;
+
+            this.contextsCache.clear();
+            const contexts = await this.contextsCache.getValue();
+            const currentContext = contexts.find(c => c.Current);
+
+            if (currentContext.DockerEndpoint === '') { // TODO: check based on type

```

Comment:
This is a little awkward. Maybe use a map (endpoint type) --> (Docker client factory method)?

I am also wondering if we should always create a new client when refreshing, even if the current client is of the expected kind.

Reply:
For ACI at least, we shouldn't need to recreate on context switch, because the SDK + daemon should be context aware.